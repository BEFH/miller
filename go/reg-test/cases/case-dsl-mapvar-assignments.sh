
# ----------------------------------------------------------------
announce MAPVAR ASSIGNMENTS TO FULL SREC

run_mlr --from $indir/xyz2 put '$* = {"a":1, "b": {"x":8,"y":9}, "c":3}'
run_mlr --from $indir/xyz2 put '$* = $*'
run_mlr --from $indir/xyz2 put '@a = 1; $* = @a'
run_mlr --from $indir/xyz2 put '@b[1] = 2; $* = @b'
run_mlr --from $indir/xyz2 put '@c[1][2] = 3; $* = @c'
run_mlr --from $indir/xyz2 put '@a = 1; $* = @*'
run_mlr --from $indir/xyz2 put '@b[1] = 2; $* = @*'
run_mlr --from $indir/xyz2 put '@c[1][2] = 3; $* = @*'
run_mlr --from $indir/xyz2 put 'a = 1; $* = a'
run_mlr --from $indir/xyz2 put 'b[1] = 2; $* = b'
run_mlr --from $indir/xyz2 put 'c[1][2] = 3; $* = c'
run_mlr --from $indir/xyz2 put '$* = 3'

run_mlr --from $indir/xyz2 put '
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  map m = map_valued_func();
  $* = m;
'

run_mlr --from $indir/xyz2 put '
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  $* = map_valued_func();
'

# ----------------------------------------------------------------
announce MAPVAR ASSIGNMENTS TO FULL OOSVAR

run_mlr --from $indir/xyz2 put '@* = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump'
run_mlr --from $indir/xyz2 put '@* = $*; dump'
run_mlr --from $indir/xyz2 put '@* = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump; @* = @*; dump'
run_mlr --from $indir/xyz2 put '@a = 1; @* = @a; dump'
run_mlr --from $indir/xyz2 put '@b[1] = 2; @* = @b; dump'
run_mlr --from $indir/xyz2 put '@c[1][2] = 3; @* = @c; dump'
run_mlr --from $indir/xyz2 put '@a = 1; @* = @*; dump'
run_mlr --from $indir/xyz2 put '@b[1] = 2; @* = @*; dump'
run_mlr --from $indir/xyz2 put '@c[1][2] = 3; @* = @*; dump'
run_mlr --from $indir/xyz2 put 'a = 1; @* = a; dump'
run_mlr --from $indir/xyz2 put 'b[1] = 2; @* = b; dump'
run_mlr --from $indir/xyz2 put 'c[1][2] = 3; @* = c; dump'
run_mlr --from $indir/xyz2 put '@* = 3'

run_mlr --from $indir/xyz2 put '
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  map m = map_valued_func();
  @* = m;
  dump
'

run_mlr --from $indir/xyz2 put '
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  @* = map_valued_func();
  dump
'

# ----------------------------------------------------------------
announce MAPVAR ASSIGNMENTS TO OOSVAR

run_mlr --from $indir/xyz2 put -q '@o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump @o'
run_mlr --from $indir/xyz2 put -q '@o = @o; dump @o'
run_mlr --from $indir/xyz2 put -q '@o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump; @o = @*; dump'
run_mlr --from $indir/xyz2 put -q '@o = $*; dump @o'
run_mlr --from $indir/xyz2 put -q '@o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump @o; @o = @o; dump @o'
run_mlr --from $indir/xyz2 put -q '@o = {"a":1, "b": {"x":8,"y":9}, "c":3}; @o = @o; dump @o'
run_mlr --from $indir/xyz2 put -q '@a = 1; @o = @a; dump @o'
run_mlr --from $indir/xyz2 put -q '@b[1] = 2; @o = @b; dump @o'
run_mlr --from $indir/xyz2 put -q '@c[1][2] = 3; @o = @c; dump @o'
run_mlr --from $indir/xyz2 put -q '@a = 1; @o = @*; dump @o'
run_mlr --from $indir/xyz2 put -q '@b[1] = 2; @o = @*; dump @o'
run_mlr --from $indir/xyz2 put -q '@c[1][2] = 3; @o = @*; dump @o'
run_mlr --from $indir/xyz2 put -q 'a = 1; @o = a; dump @o'
run_mlr --from $indir/xyz2 put -q 'b[1] = 2; @o = b; dump @o'
run_mlr --from $indir/xyz2 put -q 'c[1][2] = 3; @o = c; dump @o'
run_mlr --from $indir/xyz2 put -q 'func map_valued_func() { return {"a":1,"b":2}} @o = map_valued_func(); dump @o'

# ----------------------------------------------------------------
announce MAPVAR ASSIGNMENTS TO MAP-TYPED LOCAL

run_mlr         --from $indir/xyz2 put -q 'map o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump o'
mlr_expect_fail --from $indir/xyz2 put -q 'map o = @o; dump o'
run_mlr         --from $indir/xyz2 put -q 'map o = $*; dump o'
run_mlr         --from $indir/xyz2 put -q 'map o = {"a":1, "b": {"x":8,"y":9}, "c":3}; o = o; dump o'
mlr_expect_fail --from $indir/xyz2 put -q '@a = 1; map o = @a; dump o'
run_mlr         --from $indir/xyz2 put -q '@b[1] = 2; map o = @b; dump o'
run_mlr         --from $indir/xyz2 put -q '@c[1][2] = 3; map o = @c; dump o'
run_mlr         --from $indir/xyz2 put -q '@a = 1; map o = @*; dump o'
run_mlr         --from $indir/xyz2 put -q '@b[1] = 2; map o = @*; dump o'
run_mlr         --from $indir/xyz2 put -q '@c[1][2] = 3; map o = @*; dump o'
mlr_expect_fail --from $indir/xyz2 put -q 'a = 1; map o = a; dump o'
run_mlr         --from $indir/xyz2 put -q 'b[1] = 2; map o = b; dump o'
run_mlr         --from $indir/xyz2 put -q 'c[1][2] = 3; map o = c; dump o'
run_mlr --from $indir/xyz2 put -q 'func map_valued_func() { return {"a":1,"b":2}} map o = map_valued_func(); dump o'

# ----------------------------------------------------------------
announce MAPVAR ASSIGNMENTS TO VAR-TYPED LOCAL

run_mlr --from $indir/xyz2 put -q 'var o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump o'
run_mlr --from $indir/xyz2 put -q 'var o = @o; dump o'
run_mlr --from $indir/xyz2 put -q 'var o = $*; dump o'
run_mlr --from $indir/xyz2 put -q 'var o = {"a":1, "b": {"x":8,"y":9}, "c":3}; o = o; dump o'
run_mlr --from $indir/xyz2 put -q '@a = 1; var o = @a; dump o'
run_mlr --from $indir/xyz2 put -q '@b[1] = 2; var o = @b; dump o'
run_mlr --from $indir/xyz2 put -q '@c[1][2] = 3; var o = @c; dump o'
run_mlr --from $indir/xyz2 put -q '@a = 1; var o = @*; dump o'
run_mlr --from $indir/xyz2 put -q '@b[1] = 2; var o = @*; dump o'
run_mlr --from $indir/xyz2 put -q '@c[1][2] = 3; var o = @*; dump o'
run_mlr --from $indir/xyz2 put -q 'a = 1; var o = a; dump o'
run_mlr --from $indir/xyz2 put -q 'b[1] = 2; var o = b; dump o'
run_mlr --from $indir/xyz2 put -q 'c[1][2] = 3; var o = c; dump o'
run_mlr --from $indir/xyz2 put -q 'func map_valued_func() { return {"a":1,"b":2}} var o = map_valued_func(); dump o'

# ----------------------------------------------------------------
announce MAPVAR ASSIGNMENTS TO UNTYPED LOCAL

run_mlr --from $indir/xyz2 put -q 'o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump o'
run_mlr --from $indir/xyz2 put -q 'o = @o; dump o'
run_mlr --from $indir/xyz2 put -q 'o = $*; dump o'
run_mlr --from $indir/xyz2 put -q 'o = {"a":1, "b": {"x":8,"y":9}, "c":3}; o = o; dump o'
run_mlr --from $indir/xyz2 put -q '@a = 1; o = @a; dump o'
run_mlr --from $indir/xyz2 put -q '@b[1] = 2; o = @b; dump o'
run_mlr --from $indir/xyz2 put -q '@c[1][2] = 3; o = @c; dump o'
run_mlr --from $indir/xyz2 put -q '@a = 1; o = @*; dump o'
run_mlr --from $indir/xyz2 put -q '@b[1] = 2; o = @*; dump o'
run_mlr --from $indir/xyz2 put -q '@c[1][2] = 3; o = @*; dump o'
run_mlr --from $indir/xyz2 put -q 'a = 1; o = a; dump o'
run_mlr --from $indir/xyz2 put -q 'b[1] = 2; o = b; dump o'
run_mlr --from $indir/xyz2 put -q 'c[1][2] = 3; o = c; dump o'
run_mlr --from $indir/xyz2 put -q 'func map_valued_func() { return {"a":1,"b":2}} o = map_valued_func(); dump o'
