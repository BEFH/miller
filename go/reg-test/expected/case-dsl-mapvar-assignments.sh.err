
Cannot assign non-map to existing map; got int.
Cannot assign non-map to existing map; got int.
Cannot assign non-map to existing map; got int.
Cannot assign non-map to existing map; got int.
Cannot assign non-map to existing map; got int.
Cannot assign non-map to existing map; got int.
Miller: couldn't assign variable map o from value int 1

Miller: couldn't assign variable map o from value int 1

