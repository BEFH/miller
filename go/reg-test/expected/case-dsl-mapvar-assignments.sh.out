

================================================================
MAPVAR ASSIGNMENTS TO FULL SREC

mlr --from ./reg-test/input/xyz2 put $* = {"a":1, "b": {"x":8,"y":9}, "c":3}
a=1,b.x=8,b.y=9,c=3
a=1,b.x=8,b.y=9,c=3

mlr --from ./reg-test/input/xyz2 put $* = $*
x=3,y=4,z=5
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @a = 1; $* = @a

mlr --from ./reg-test/input/xyz2 put @b[1] = 2; $* = @b
1=2
1=2

mlr --from ./reg-test/input/xyz2 put @c[1][2] = 3; $* = @c
1.2=3
1.2=3

mlr --from ./reg-test/input/xyz2 put @a = 1; $* = @*
a=1
a=1

mlr --from ./reg-test/input/xyz2 put @b[1] = 2; $* = @*
b.1=2
b.1=2

mlr --from ./reg-test/input/xyz2 put @c[1][2] = 3; $* = @*
c.1.2=3
c.1.2=3

mlr --from ./reg-test/input/xyz2 put a = 1; $* = a

mlr --from ./reg-test/input/xyz2 put b[1] = 2; $* = b
1=2
1=2

mlr --from ./reg-test/input/xyz2 put c[1][2] = 3; $* = c
1.2=3
1.2=3

mlr --from ./reg-test/input/xyz2 put $* = 3

mlr --from ./reg-test/input/xyz2 put 
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  map m = map_valued_func();
  $* = m;

a=1,b=2
a=1,b=2

mlr --from ./reg-test/input/xyz2 put 
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  $* = map_valued_func();

a=1,b=2
a=1,b=2


================================================================
MAPVAR ASSIGNMENTS TO FULL OOSVAR

mlr --from ./reg-test/input/xyz2 put @* = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
x=3,y=4,z=5
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @* = $*; dump
{
  "x": 3,
  "y": 4,
  "z": 5
}
x=3,y=4,z=5
{
  "x": 6,
  "y": 7,
  "u": 8
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @* = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump; @* = @*; dump
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
x=3,y=4,z=5
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @a = 1; @* = @a; dump

mlr --from ./reg-test/input/xyz2 put @b[1] = 2; @* = @b; dump
{
  "1": 2
}
x=3,y=4,z=5
{
  "1": 2
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @c[1][2] = 3; @* = @c; dump
{
  "1": {
    "2": 3
  }
}
x=3,y=4,z=5
{
  "1": {
    "2": 3
  }
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @a = 1; @* = @*; dump
{
  "a": 1
}
x=3,y=4,z=5
{
  "a": 1
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @b[1] = 2; @* = @*; dump
{
  "b": {
    "1": 2
  }
}
x=3,y=4,z=5
{
  "b": {
    "1": 2
  }
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @c[1][2] = 3; @* = @*; dump
{
  "c": {
    "1": {
      "2": 3
    }
  }
}
x=3,y=4,z=5
{
  "c": {
    "1": {
      "2": 3
    }
  }
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put a = 1; @* = a; dump

mlr --from ./reg-test/input/xyz2 put b[1] = 2; @* = b; dump
{
  "1": 2
}
x=3,y=4,z=5
{
  "1": 2
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put c[1][2] = 3; @* = c; dump
{
  "1": {
    "2": 3
  }
}
x=3,y=4,z=5
{
  "1": {
    "2": 3
  }
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put @* = 3

mlr --from ./reg-test/input/xyz2 put 
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  map m = map_valued_func();
  @* = m;
  dump

{
  "a": 1,
  "b": 2
}
x=3,y=4,z=5
{
  "a": 1,
  "b": 2
}
x=6,y=7,u=8

mlr --from ./reg-test/input/xyz2 put 
  func map_valued_func() {
    return {"a":1,"b":2}
  }
  @* = map_valued_func();
  dump

{
  "a": 1,
  "b": 2
}
x=3,y=4,z=5
{
  "a": 1,
  "b": 2
}
x=6,y=7,u=8


================================================================
MAPVAR ASSIGNMENTS TO OOSVAR

mlr --from ./reg-test/input/xyz2 put -q @o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump @o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q @o = @o; dump @o

mlr --from ./reg-test/input/xyz2 put -q @o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump; @o = @*; dump
{
  "o": {
    "a": 1,
    "b": {
      "x": 8,
      "y": 9
    },
    "c": 3
  }
}
{
  "o": {
    "o": {
      "a": 1,
      "b": {
        "x": 8,
        "y": 9
      },
      "c": 3
    }
  }
}
{
  "o": {
    "a": 1,
    "b": {
      "x": 8,
      "y": 9
    },
    "c": 3
  }
}
{
  "o": {
    "o": {
      "a": 1,
      "b": {
        "x": 8,
        "y": 9
      },
      "c": 3
    }
  }
}

mlr --from ./reg-test/input/xyz2 put -q @o = $*; dump @o
{
  "x": 3,
  "y": 4,
  "z": 5
}
{
  "x": 6,
  "y": 7,
  "u": 8
}

mlr --from ./reg-test/input/xyz2 put -q @o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump @o; @o = @o; dump @o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q @o = {"a":1, "b": {"x":8,"y":9}, "c":3}; @o = @o; dump @o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; @o = @a; dump @o
1
1

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; @o = @b; dump @o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; @o = @c; dump @o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; @o = @*; dump @o
{
  "a": 1
}
{
  "a": 1,
  "o": {
    "a": 1
  }
}

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; @o = @*; dump @o
{
  "b": {
    "1": 2
  }
}
{
  "b": {
    "1": 2
  },
  "o": {
    "b": {
      "1": 2
    }
  }
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; @o = @*; dump @o
{
  "c": {
    "1": {
      "2": 3
    }
  }
}
{
  "c": {
    "1": {
      "2": 3
    }
  },
  "o": {
    "c": {
      "1": {
        "2": 3
      }
    }
  }
}

mlr --from ./reg-test/input/xyz2 put -q a = 1; @o = a; dump @o
1
1

mlr --from ./reg-test/input/xyz2 put -q b[1] = 2; @o = b; dump @o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q c[1][2] = 3; @o = c; dump @o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q func map_valued_func() { return {"a":1,"b":2}} @o = map_valued_func(); dump @o
{
  "a": 1,
  "b": 2
}
{
  "a": 1,
  "b": 2
}


================================================================
MAPVAR ASSIGNMENTS TO MAP-TYPED LOCAL

mlr --from ./reg-test/input/xyz2 put -q map o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q map o = @o; dump o

mlr --from ./reg-test/input/xyz2 put -q map o = $*; dump o
{
  "x": 3,
  "y": 4,
  "z": 5
}
{
  "x": 6,
  "y": 7,
  "u": 8
}

mlr --from ./reg-test/input/xyz2 put -q map o = {"a":1, "b": {"x":8,"y":9}, "c":3}; o = o; dump o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; map o = @a; dump o

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; map o = @b; dump o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; map o = @c; dump o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; map o = @*; dump o
{
  "a": 1
}
{
  "a": 1
}

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; map o = @*; dump o
{
  "b": {
    "1": 2
  }
}
{
  "b": {
    "1": 2
  }
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; map o = @*; dump o
{
  "c": {
    "1": {
      "2": 3
    }
  }
}
{
  "c": {
    "1": {
      "2": 3
    }
  }
}

mlr --from ./reg-test/input/xyz2 put -q a = 1; map o = a; dump o

mlr --from ./reg-test/input/xyz2 put -q b[1] = 2; map o = b; dump o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q c[1][2] = 3; map o = c; dump o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q func map_valued_func() { return {"a":1,"b":2}} map o = map_valued_func(); dump o
{
  "a": 1,
  "b": 2
}
{
  "a": 1,
  "b": 2
}


================================================================
MAPVAR ASSIGNMENTS TO VAR-TYPED LOCAL

mlr --from ./reg-test/input/xyz2 put -q var o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q var o = @o; dump o

mlr --from ./reg-test/input/xyz2 put -q var o = $*; dump o
{
  "x": 3,
  "y": 4,
  "z": 5
}
{
  "x": 6,
  "y": 7,
  "u": 8
}

mlr --from ./reg-test/input/xyz2 put -q var o = {"a":1, "b": {"x":8,"y":9}, "c":3}; o = o; dump o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; var o = @a; dump o
1
1

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; var o = @b; dump o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; var o = @c; dump o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; var o = @*; dump o
{
  "a": 1
}
{
  "a": 1
}

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; var o = @*; dump o
{
  "b": {
    "1": 2
  }
}
{
  "b": {
    "1": 2
  }
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; var o = @*; dump o
{
  "c": {
    "1": {
      "2": 3
    }
  }
}
{
  "c": {
    "1": {
      "2": 3
    }
  }
}

mlr --from ./reg-test/input/xyz2 put -q a = 1; var o = a; dump o
1
1

mlr --from ./reg-test/input/xyz2 put -q b[1] = 2; var o = b; dump o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q c[1][2] = 3; var o = c; dump o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q func map_valued_func() { return {"a":1,"b":2}} var o = map_valued_func(); dump o
{
  "a": 1,
  "b": 2
}
{
  "a": 1,
  "b": 2
}


================================================================
MAPVAR ASSIGNMENTS TO UNTYPED LOCAL

mlr --from ./reg-test/input/xyz2 put -q o = {"a":1, "b": {"x":8,"y":9}, "c":3}; dump o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q o = @o; dump o

mlr --from ./reg-test/input/xyz2 put -q o = $*; dump o
{
  "x": 3,
  "y": 4,
  "z": 5
}
{
  "x": 6,
  "y": 7,
  "u": 8
}

mlr --from ./reg-test/input/xyz2 put -q o = {"a":1, "b": {"x":8,"y":9}, "c":3}; o = o; dump o
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}
{
  "a": 1,
  "b": {
    "x": 8,
    "y": 9
  },
  "c": 3
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; o = @a; dump o
1
1

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; o = @b; dump o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; o = @c; dump o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q @a = 1; o = @*; dump o
{
  "a": 1
}
{
  "a": 1
}

mlr --from ./reg-test/input/xyz2 put -q @b[1] = 2; o = @*; dump o
{
  "b": {
    "1": 2
  }
}
{
  "b": {
    "1": 2
  }
}

mlr --from ./reg-test/input/xyz2 put -q @c[1][2] = 3; o = @*; dump o
{
  "c": {
    "1": {
      "2": 3
    }
  }
}
{
  "c": {
    "1": {
      "2": 3
    }
  }
}

mlr --from ./reg-test/input/xyz2 put -q a = 1; o = a; dump o
1
1

mlr --from ./reg-test/input/xyz2 put -q b[1] = 2; o = b; dump o
{
  "1": 2
}
{
  "1": 2
}

mlr --from ./reg-test/input/xyz2 put -q c[1][2] = 3; o = c; dump o
{
  "1": {
    "2": 3
  }
}
{
  "1": {
    "2": 3
  }
}

mlr --from ./reg-test/input/xyz2 put -q func map_valued_func() { return {"a":1,"b":2}} o = map_valued_func(); dump o
{
  "a": 1,
  "b": 2
}
{
  "a": 1,
  "b": 2
}

